buffer/src/main/java/io/netty5/buffer/api/DefaultCompositeBuffer.java

```diff
@@ -160,53 +159,162 @@ public static CompositeBuffer compose(BufferAllocator allocator) {
-        Buffer[] bufs = refs
-                .filter(DefaultCompositeBuffer::discardEmpty)
-                .flatMap(DefaultCompositeBuffer::flattenBuffer)
-                .toArray(Buffer[]::new);
-        // Make sure there are no duplicates among the buffers.
-        Set<Buffer> duplicatesCheck = Collections.newSetFromMap(new IdentityHashMap<>());
-        duplicatesCheck.addAll(Arrays.asList(bufs));
-        if (duplicatesCheck.size() < bufs.length) {
-            IllegalArgumentException iae = new IllegalArgumentException(
-                    "Cannot create composite buffer with duplicate constituent buffer components.");
-            for (Buffer buf : bufs) {
+        Collector collector = new Collector(externals);
+        collector.collect(externals);
+        return collector.toArray();
+    }
+
+    private static final class Collector {
+        Buffer[] array;
+        int index;
+
+        Collector(Iterable<Buffer> externals) {
+            var dupeCheck = new IdentityHashMap<>();
+            int size = 0;
+            for (Buffer buf : externals) {
+                if (dupeCheck.put(buf, buf) != null) {
+                    // Throw if there are duplicates among the buffers.
+                    // We can do this before we decompose composites, because the components are owned,
+                    // and no two composite buffers can have the same components, unless they violate
+                    // their API contract.
+                    closeAllAndThrowDupeException(externals);
+                }
+                size += buf.countComponents();
+            }
+            array = new Buffer[size];
+        }
+
+        private static void closeAllAndThrowDupeException(Iterable<Buffer> externals) {
+            IllegalArgumentException iae = new IllegalArgumentException("Cannot compose duplicate buffers.");
+            for (Buffer toClose : externals) {
                 try {
-                    buf.close();
+                    toClose.close();
                 } catch (Exception closeExc) {
                     iae.addSuppressed(closeExc);
                 }
             }
             throw iae;
         }
-        return bufs;
+
+        void add(Buffer buffer) {
+            if (index == array.length) {
+                array = Arrays.copyOf(array, array.length * 2);
+            }
+            array[index] = buffer;
+            index++;
+        }
+
+        void collect(Iterable<Buffer> externals) {
+            for (Buffer buf : externals) {
+                if (buf.capacity() == 0) {
+                    buf.close();
+                    continue;
+                }
+                if (CompositeBuffer.isComposite(buf)) {
+                    CompositeBuffer cbuf = (CompositeBuffer) buf;
+                    collect(Arrays.asList(cbuf.decomposeBuffer()));
+                    continue;
+                }
+                add(buf);
+            }
```

Comment:
Nit: I think this is a bit easier to read:


```suggestion
            for (Buffer buf : externals) {
                if (buf.capacity() == 0) {
                    buf.close();
                } else  if (CompositeBuffer.isComposite(buf)) {
                    CompositeBuffer cbuf = (CompositeBuffer) buf;
                    collect(Arrays.asList(cbuf.decomposeBuffer()));
                } else {
                    add(buf);
                }
            }
```